### PR TITLE
feat!: Allow linear data inside lists

### LINKED ISSUE #524: Use inout for lists

### LINKED ISSUE #530: Allow linear types inside lists

### PR DESCRIPTION
**Pull request by mark-koch**
Closes #530 and closes #524.

* Changes the signature of list methods to allow linear list elements (implicitly using the inout feature).
* Removes methods that are not supported for now (see #528)
* Small change to the unification logic to be a bit laxer about requiring matching `@owned` flags if one of the arguments is not linear (otherwise we couldn't unify `L @owned` with any classical type). We'll need to revisit that once we have flags that can be put on classical types.

BREAKING CHANGE: Unsupported list methods have been removed.

### CODE REVIEW

File: tests/error/poly_errors/non_linear2.err [hunk 1]
```diff
@@ -3,5 +3,5 @@ Guppy compilation failed. Error in file $FILE:23
 21:    @guppy(module)
 22:    def main() -> None:
 23:        foo(h)
-               ^
-GuppyTypeError: Expected argument of type `?T -> ?T`, got `qubit @owned -> qubit`
+           ^^^^^^
+GuppyTypeError: Cannot instantiate non-linear type variable `T` in type `forall T. (T -> T) -> None` with linear type `qubit`
```
Comment (mark-koch) [inline tests/error/poly_errors/non_linear2.err:7]: A side effect of the unification change is that we get a better error message here.

However, I think #487 still applies

File: guppylang/prelude/builtins.py [hunk 2]
```diff
@@ -505,77 +504,35 @@ def __trunc__(self: float) -> float: ...
...
 
-    @guppy.hugr_op(unsupported_op("IsNotEmpty"))
-    def __hasnext__(self: list[T]) -> tuple[bool, list[T]]: ...
-
-    @guppy.custom(NoopCompiler())
-    def __iter__(self: list[T]) -> list[T]: ...
-
-    @guppy.hugr_op(unsupported_op("Length"))
-    def __len__(self: list[T]) -> int: ...
+    @guppy.hugr_op(unsupported_op("pop"))  # TODO: check None and unwrap
```
Comment (ss2165) [inline guppylang/prelude/builtins.py:510]: "pop" doesn't seem right here?

File: guppylang/prelude/builtins.py [hunk 3]
```diff
@@ -505,77 +504,35 @@ def __trunc__(self: float) -> float: ...
...
+    def __setitem__(self: list[L], idx: int, value: L @ owned) -> None: ...
 
-    @guppy.hugr_op(unsupported_op("Repeat"))
-    def __mul__(self: list[T], other: int) -> list[T]: ...
-
-    @guppy.hugr_op(unsupported_op("Pop"))
-    def __next__(self: list[T]) -> tuple[T, list[T]]: ...
+    @guppy.hugr_op(unsupported_op("length"))  # TODO: inout return in wrong order
```
Comment (ss2165) [inline guppylang/prelude/builtins.py:513]: is this ordering being resolved elsewhere?

File: guppylang/prelude/builtins.py [hunk 4]
```diff
@@ -505,77 +504,35 @@ def __trunc__(self: float) -> float: ...
...
-
-    @guppy.hugr_op(unsupported_op("Find"))
-    def index(self: list[T], elt: T) -> int: ...
-
-    @guppy.custom(checker=FailingChecker("Guppy lists are immutable"))
-    def pop(self: list[T], idx: int) -> None: ...
-
-    @guppy.custom(checker=FailingChecker("Guppy lists are immutable"))
-    def remove(self: list[T], elt: T) -> None: ...
+    @guppy.hugr_op(unsupported_op("Pop"))
```
Comment (ss2165) [inline guppylang/prelude/builtins.py:528]: case difference between "pop" and "Pop" ?

**Review by ss2165 — COMMENTED** [review]
Looks good, few queries on the unsupporteds

File: guppylang/prelude/builtins.py (hunk 3 above)
Comment (mark-koch) [reply guppylang/prelude/builtins.py:513]: Actually turns out that the length op doesn't return the list at all! https://github.com/CQCL/hugr/issues/1543

**Review by ss2165 — APPROVED** [review]

**Pull request merged**